# Лабораторная работа 8
- **Дедлайн**: 11.06.2023, 23:59
- **Максимум баллов**: 10

В первой главе данного курса вам было предложено с помощью фреймворка sklearn обучить модель машинного обучения для задачи бинарной классификации на наборе данных Titanic. Давайте попробуем выполнить ту же самую задачу с помощью фреймворка глубокого обучения TensorFlow.

Давайте убедимся, что необходимые нам для этой работы библиотеки уже установлены.

In [1]:
!pip install numpy tensorflow_datasets tensorflow

Defaulting to user installation because normal site-packages is not writeable


# Описание набора данных
В данной работе целью исследования будет набор данных Titanic. Данный набор содержит информацию о различных пассажирах корабля "Титаник", затонувшего в ночь на 15-е апреля 1912 года. Некоторое количество пассажиров спаслось, чему способствовало множество различных факторов, включая их пол, возраст, на какой палубе находилась их кабина, социальный статус, и т.д. Мы предлагаем вам натренировать нейронную сеть для бинарной классификации, способную предсказывать вероятность спасения человека на основе его данных.  

Набор данных состоит из различных признаков, описывающих информацию о пассажирах. Каждая строка таблицы - отдельный пассажир, вся информация о нем содержится в его строке. В столбце survived находится бинарная метка (0 или 1), означающая, спасся ли человек с корабля (1) или нет (0). Вашей задачей является изучить набор данных, корректно определить решаемую задачу, а затем построить модель, которая будет способна предсказывать вероятность спасения и посчитать F1-метрику на тестовом наборе данных.

Для загрузки набора данных мы снова будем использовать библиотеку tensorflow_datasets. 

Импортируем данную библиотеку, а затем с помощью нее загрузим набор данных "titanic". В процессе загрузки сразу разделим его на тренировочную, валидационную и тестовую выборку.

In [2]:
import tensorflow_datasets as tfds
ds_train_tf, ds_validation_tf, ds_test_tf = tfds.load(
    name='titanic',
    split=['train[:80%]', 'train[80%:90%]', 'train[90%:]'],
    as_supervised=True
)

По умолчанию библиотека загружает данные в формате TensorFlow Dataset. Этот формат позволяет оперировать данными на жестком диске без предварительной загрузки их в память, что позволят обучать модели на данных, превышающих размер оперативной памяти вашего устройства.

Набор данных Titanic является достаточно небольшим набором и мы уверены, что оперативной памяти вашего устройства хватит для хранения его целиком, поэтому это в данный момент не играет решающей роли. Однако, вам будет полезно научиться работать с ним.

Примеры использования и различных функций этого формата данных вы можете найти в документации или Jupyter Notebook'е данного курса, описывающем работу с библиотекой Keras.

Давайте выведем первую строку тренировочного набора данных для ознакомления.

In [3]:
for x, y in ds_train_tf.take(1):
    print(f"{x=}")
    print(f"{y=}")

x={'age': <tf.Tensor: shape=(), dtype=float32, numpy=30.0>, 'boat': <tf.Tensor: shape=(), dtype=string, numpy=b'Unknown'>, 'body': <tf.Tensor: shape=(), dtype=int32, numpy=-1>, 'cabin': <tf.Tensor: shape=(), dtype=string, numpy=b'Unknown'>, 'embarked': <tf.Tensor: shape=(), dtype=int64, numpy=2>, 'fare': <tf.Tensor: shape=(), dtype=float32, numpy=13.0>, 'home.dest': <tf.Tensor: shape=(), dtype=string, numpy=b'Sarnia, ON'>, 'name': <tf.Tensor: shape=(), dtype=string, numpy=b'McCrie, Mr. James Matthew'>, 'parch': <tf.Tensor: shape=(), dtype=int32, numpy=0>, 'pclass': <tf.Tensor: shape=(), dtype=int64, numpy=1>, 'sex': <tf.Tensor: shape=(), dtype=int64, numpy=0>, 'sibsp': <tf.Tensor: shape=(), dtype=int32, numpy=0>, 'ticket': <tf.Tensor: shape=(), dtype=string, numpy=b'233478'>}
y=<tf.Tensor: shape=(), dtype=int64, numpy=0>


Как можно заметить, формат текущего датасета представляет собой кортеж из двух элементов - признаков и метки класса. Признаки организованы как словарь с ключами (названиями признаков) и значениями признаков. Данный формат можно использовать напрямую для метода model.fit, однако признаки, несомненно, требуют предобработки.

Итак, к настоящему моменту мы загрузили набор данных, разделили его на тренировочную, валидационную и тестовую выборку и готовы приступить к решению задачи. Вашим финальным заданием является получение максимального значения F1-метрики на тестовом наборе данных, используя все возможности по изучению и визуализации данных, предобработке данных, а также нейронную сеть, выполненную с помощью фреймворка TensorFlow (и его части - Keras).  

Для того, чтобы напомнить процесс подготовки данных и обучения модели, мы поместили несколько ячеек с комментариями ниже в предполагаемом порядке. Однако, вы можете использовать любой удобный вам процесс.  

В процессе исследования и тренировки запрещено использовать данных из ds_test кроме как для финальной оценки F1 метрики. Данная лабораторная работа служит не для проверки ваших навыков, а для того, чтобы помочь вам вспомнить основы машинного обучения и работы с соответствующими библиотеками. Мы просим вас попытаться решить задачу самостоятельно, не используя доступную информацию из ds_test набора и не пытаясь подогнать модель под идеальную работу с конкретно данным набором, а также просим не использовать внешние источники данных и информацию из соответствующих соревнований на платформе Kaggle.

Вы можете найти для себя полезным пройти снова ноутбук с описанием работы фреймворка Keras и информацией о том, как его использовать. Также по данному фреймворку доступны многочисленные туториалы и примеры на его сайте. Напомним вам, что задача представляет собой бинарную классификацию, поэтому вам необходимо выбрать правильное количество нейронов для последнего слоя и функцию активации.

Успехов!

# Предобработка данных

* 1. Предобработка данных: заполнение пропущенных мест, приведение столбцов к единому виду, удаление nan/infinity значений
* 2. Инженерия признаков: генерация или выделение наиболее интересных признаков, удаление ненужных
* 3. Нормализация данных относительно среднего и медианы

In [4]:
import tensorflow as tf

# Define the preprocessing function
def preprocess(features, label):
    # Fill in missing values
    age_mean = tf.reduce_mean(features['age'])
    fare_mean = tf.reduce_mean(features['fare'])
    features['age'] = tf.where(tf.math.is_nan(features['age']), age_mean, features['age'])
    features['fare'] = tf.where(tf.math.is_nan(features['fare']), fare_mean, features['fare'])
    
    # Create new feature: total family members on board
    features['family_members'] = features['sibsp'] + features['parch']
    
    # One-hot encode categorical features
    sex = tf.one_hot(tf.cast(tf.as_string(features['sex']) == 'male', tf.int32), depth=2)
    pclass = tf.one_hot(features['pclass'] - 1, depth=3)
    
    # Extract title from name and one-hot encode it
    title = tf.strings.split(features['name'], ',')[1]
    title = tf.strings.split(title)[0]
    title = tf.where(title == 'Miss.', 'Ms.', title)
    title = tf.where(title == 'Mlle.', 'Ms.', title)
    title = tf.where(title == 'Mme.', 'Mrs.', title)
    unique_titles = ['Mr.', 'Mrs.', 'Ms.', 'Master.', 'Dr.', 'Rev.']
    title = tf.one_hot(tf.argmax(tf.stack([tf.cast(title == t, tf.int32) for t in unique_titles])), depth=len(unique_titles))
    
    # Normalize numerical features
    epsilon = 1e-8
    age = (features['age'] - age_mean) / (tf.math.reduce_std(features['age']) + epsilon)
    fare = (features['fare'] - fare_mean) / (tf.math.reduce_std(features['fare']) + epsilon)
    
    # Expand dimensions of scalar features
    age = tf.expand_dims(age, axis=-1)
    fare = tf.expand_dims(fare, axis=-1)
    
    # Concatenate all features into a single tensor
    inputs = tf.concat([sex, pclass, title, age, fare], axis=-1)
    
    return inputs, label

In [5]:
# Apply the preprocessing function to the data
ds_train_tf = ds_train_tf.map(preprocess)
ds_validation_tf = ds_validation_tf.map(preprocess)
ds_test_tf = ds_test_tf.map(preprocess)

In [6]:
# View the first few examples in the training dataset
for features, label in ds_train_tf.take(1):
    print(f'Features: {features}')
    print(f'Label: {label}')
    print()

Features: [1. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
Label: 0



# Обучение модели

* 1. Выбор модели или ансамбля моделей
* 2. Поиск оптимальных гиперпараметров на validation наборе данных

In [7]:
# Define the model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(62, activation='relu'),
#     tf.keras.layers.Dense(32, activation='relu'),
#     tf.keras.layers.Dense(16, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

In [8]:
learning_rates = [0.1, 0.01, 0.001]
num_units = [32, 64, 128]

best_val_acc = 0
best_lr = None
best_units = None

for lr in learning_rates:
    for units in num_units:
        # Define the model
        model = tf.keras.Sequential([
            tf.keras.layers.Dense(units, activation='relu'),
            tf.keras.layers.Dense(1, activation='sigmoid')
        ])
        # Compile the model
        optimizer = tf.keras.optimizers.Adam(learning_rate=lr)
        model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])
        # Train the model
        history = model.fit(ds_train_tf.batch(32), validation_data=ds_validation_tf.batch(32), epochs=10)
        # Get the best validation accuracy
        val_acc = max(history.history['val_accuracy'])
        # Update the best hyperparameters if necessary
        if val_acc > best_val_acc:
            best_val_acc = val_acc
            best_lr = lr
            best_units = units

print(f'Best validation accuracy: {best_val_acc}')
print(f'Best learning rate: {best_lr}')
print(f'Best number of units: {best_units}')

Epoch 1/10
33/33 [==============================] - 1s 8ms/step - loss: 0.5262 - accuracy: 0.7564 - val_loss: 0.4069 - val_accuracy: 0.8321
Epoch 2/10
33/33 [==============================] - 0s 4ms/step - loss: 0.4670 - accuracy: 0.7870 - val_loss: 0.3995 - val_accuracy: 0.8168
Epoch 3/10
33/33 [==============================] - 0s 3ms/step - loss: 0.4651 - accuracy: 0.7861 - val_loss: 0.4016 - val_accuracy: 0.8092
Epoch 4/10
33/33 [==============================] - 0s 4ms/step - loss: 0.4630 - accuracy: 0.7908 - val_loss: 0.3930 - val_accuracy: 0.8321
Epoch 5/10
33/33 [==============================] - 0s 4ms/step - loss: 0.4640 - accuracy: 0.7841 - val_loss: 0.3949 - val_accuracy: 0.8321
Epoch 6/10
33/33 [==============================] - 0s 4ms/step - loss: 0.4636 - accuracy: 0.7803 - val_loss: 0.3982 - val_accuracy: 0.8397
Epoch 7/10
33/33 [==============================] - 0s 3ms/step - loss: 0.4630 - accuracy: 0.7775 - val_loss: 0.4025 - val_accuracy: 0.8092
Epoch 8/10
33/33 [==

33/33 [==============================] - 0s 3ms/step - loss: 0.4521 - accuracy: 0.7870 - val_loss: 0.4049 - val_accuracy: 0.8397
Epoch 10/10
33/33 [==============================] - 0s 3ms/step - loss: 0.4520 - accuracy: 0.7880 - val_loss: 0.4049 - val_accuracy: 0.8397
Epoch 1/10
33/33 [==============================] - 0s 6ms/step - loss: 0.6224 - accuracy: 0.7106 - val_loss: 0.6250 - val_accuracy: 0.6870
Epoch 2/10
33/33 [==============================] - 0s 3ms/step - loss: 0.5752 - accuracy: 0.7287 - val_loss: 0.5866 - val_accuracy: 0.6947
Epoch 3/10
33/33 [==============================] - 0s 3ms/step - loss: 0.5435 - accuracy: 0.7545 - val_loss: 0.5445 - val_accuracy: 0.8092
Epoch 4/10
33/33 [==============================] - 0s 3ms/step - loss: 0.5189 - accuracy: 0.7832 - val_loss: 0.5116 - val_accuracy: 0.8321
Epoch 5/10
33/33 [==============================] - 0s 3ms/step - loss: 0.5014 - accuracy: 0.7775 - val_loss: 0.4862 - val_accuracy: 0.8321
Epoch 6/10
33/33 [============

# Оценка модели

* 1. Получение F1-метрики на тестовом наборе данных

In [9]:
# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(ds_test_tf.batch(32))
print(f'Test accuracy: {test_acc}')

5/5 [==============================] - 0s 1ms/step - loss: 0.4468 - accuracy: 0.8015
Test accuracy: 0.8015267252922058


In [10]:
from sklearn.metrics import f1_score
# Calculate and print out the F1 score on the test set
y_true = tf.stack([y for x, y in ds_test_tf], axis=0).numpy()
y_pred = tf.round(model.predict(ds_test_tf.batch(32))).numpy()
test_f1 = f1_score(y_true, y_pred)
print(f'Test F1 score: {test_f1:.4f}')

5/5 [==============================] - 0s 1000us/step
Test F1 score: 0.7292


# Conclusions
* The model is working well.
* We got good results from f1, but it is still not enough.
* As we can see the accurary is not increasing fron 0.7832 which means the model has reached its maximum capacity.
* Another maybe there are still noise in the dataset or contains errors which why the model is accuracy is not increasing.
* Another possible reason for the model's accuracy not increasing is overfitting.